### PR TITLE
Fix duplicate saving in Paper Mario

### PR DESCRIPTION
**Pull request by loganmc10**
This fixes https://github.com/mupen64plus/mupen64plus-core/issues/154

Right now, if the PI controller gets a strange request, we basically just ignore it and hope for better luck next time.

What this change does: If the rom address is too large, or if the rom address is good, but the (rom_address + length) would put us outside the ROM's size, then we copy what fits, and fill the rest with 0's.

I'll do some more testing in the next day or two to make sure this doesn't cause any regressions in other games.

### CODE REVIEW

**Comment by loganmc10** [comment]
My testing of this PR hasn't revealed any issues with any other games

**Comment by Narann** [comment]
Good job @loganmc10 !

This is not a tiny change, I would wait for @bsmiles32 and @Gillou68310 to check this before merge.

**Comment by bsmiles32** [comment]
Not sure what is the real behavior on the console, but your changes seems ok. I tested with Paper Mario, did not notice anything strange, but also I wasn't able to reproduce the issue before, so I will just trust you that it indeed fixes the saving issue.

Merging this. Good work !

**Comment by loganmc10** [comment]
@bsmiles32 somebody in the libretro thread reported that it doesn't happen in the PAL version, I'm assuming you were probably testing that version.

**Comment by bsmiles32** [comment]
Indeed :)

File: src/device/pi/pi_controller.c [hunk 1]
```diff
@@ -116,47 +116,32 @@ static void dma_pi_write(struct pi_controller* pi)
...
     {
-        dram[(dram_address+i)^S8] = rom[(rom_address+i)^S8];
+        int32_t diff = pi->cart_rom.rom_size - rom_address;
+        if (diff < 0) diff = 0;
+
+        for (i = 0; i < diff; ++i)
+        {
+            dram[(dram_address+i)^S8] = rom[(rom_address+i)^S8];
+        }
+        for (i = diff; i < longueur - diff; ++i)
```
Comment (richard42) [inline src/device/pi/pi_controller.c]: I think this should be `for (; i < longueur; ++i)`

Comment (loganmc10) [reply src/device/pi/pi_controller.c]: hm I think you are right, I don't know what I was envisioned when I wrote that. I tested Paper Mario and it's still fixed. Do you want me to submit another PR?

Comment (richard42) [reply src/device/pi/pi_controller.c]: no it's okay I'll just push a fix from here.

**Comment by Fad1000** [comment]
> This fixes #154
> 
> Right now, if the PI controller gets a strange request, we basically just ignore it and hope for better luck next time.
> 
> What this change does: If the rom address is too large, or if the rom address is good, but the (rom_address + length) would put us outside the ROM's size, then we copy what fits, and fill the rest with 0's.
> 
> I'll do some more testing in the next day or two to make sure this doesn't cause any regressions in other games.

What am I supposed to do? #

**Pull request merged**